### PR TITLE
Add gRPC integration as gRPC client interceptor

### PR DESCRIPTION
**Pull request by a-simeshin**
### Context
Add a client interceptor that can attach gRPC integration data to an Allure report.

Due to the lack of documentation for java grpc-api, docs in official blog for the interceptor is on go lang, but the idea is clear - https://grpc.io/blog/grpc-web-interceptor/
JavaDocs from java grpc-api - https://grpc.github.io/grpc-java/javadoc/io/grpc/ClientInterceptor.html

The interceptor must support:
- [x] https://github.com/grpc/grpc-java
- [x] https://github.com/yidongnan/grpc-spring-boot-starter

#### Checklist
- [x] [Sign Allure CLA][cla]
- [x] Provide unit tests

[cla]: https://cla-assistant.io/accept/allure-framework/allure2

### CODE REVIEW

File: allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+
+    private static final String TEXT_PLAIN = "text/plain";
+    private final AtomicReference<A> responseContainer = new AtomicReference<>(null);
+    private final AtomicReference<Metadata> headersContainer = new AtomicReference<>(null);
+    private final AtomicReference<Status> statusContainer = new AtomicReference<>(null);
+
+    private AllureLifecycle allureLifecycle;
+    private MethodDescriptor<Q, A> methodDescriptor;
+
+    public CustomForwardingClientCall<Q, A> setAllureLifeCycle(final AllureLifecycle allureLifecycle) {
```
Comment (baev) [inline allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java:61]: let's use `setLifecycle` instead to align with other modules

File: allure-grpc/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+plugins {
+    id("com.google.protobuf") version "0.8.16"
+}
+
+dependencies {
+    agent("org.aspectj:aspectjweaver")
+    api(project(":allure-attachments"))
+    implementation("io.grpc:grpc-api")
+    implementation("org.awaitility:awaitility")
```
Comment (baev) [inline allure-grpc/build.gradle.kts:20]: it's better to have no additional dependencies

File: allure-grpc/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    testImplementation("io.grpc:grpc-protobuf")
+    testImplementation("io.grpc:grpc-stub")
+    testImplementation("org.grpcmock:grpcmock-junit5")
+    testImplementation("org.assertj:assertj-core")
+    testImplementation("org.jboss.resteasy:resteasy-client")
+    testImplementation("org.junit.jupiter:junit-jupiter-api")
+    testImplementation("org.junit.jupiter:junit-jupiter-params")
+    testImplementation("org.mockito:mockito-core")
+    testImplementation("org.slf4j:slf4j-simple")
+    testImplementation("org.slf4j:slf4j-simple")
```
Comment (baev) [inline allure-grpc/build.gradle.kts:30]: duplicate row

File: allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+    }
+
+    @Override
+    public void sendMessage(final Q message) {
+        super.sendMessage(message);
+
+        Allure.setLifecycle(allureLifecycle);
+        Allure.step("gRPC interaction " + methodDescriptor.getFullMethodName(), () -> {
+            Allure.addAttachment("gRPC request", ObjectUtils.toString(message));
+            Allure.addByteAttachmentAsync("gRPC response", TEXT_PLAIN, () -> {
```
Comment (baev) [inline allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java:82]: you should always wait for `addByteAttachmentAsync` future completion or you may lose some data. Is `addByteAttachmentAsync` required?

**Review by baev — CHANGES_REQUESTED** [review]

File: allure-grpc/build.gradle.kts (hunk 2 above)
Comment (a-simeshin) [reply allure-grpc/build.gradle.kts:20]: No way to drop grpc-api. The library provides an client interceptor interface and client calls, also much more stuff.

If it's about awaititlity, then the library provides a very convenient way to wait for events. I understand that this is not the best option, add a dependency due to 3 expectations in the code. 
I would be glad to get an advise how I can replace this with the current libraries in the project.

File: allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java (hunk 1 above)
Comment (a-simeshin) [reply allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java]: Done

File: allure-grpc/build.gradle.kts (hunk 3 above)
Comment (a-simeshin) [reply allure-grpc/build.gradle.kts]: Fixed

File: allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java (hunk 4 above)
Comment (a-simeshin) [reply allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java:87]: I completely agree that synchronous saving is much better than asynchronous, but in a real project I could not find another way to attach data. If addAttachments is called in asynronic (onMessage, onHeaders, onClose) methods, the data will be created, but not attached to the TestCaseResult. 

I was never able to figure out why this is happening, so I tried to work with an asynchronous stream in the StreamObserver part which the responseListener belongs as a synchronous one.

**Comment by a-simeshin** [comment]
@baev 

A small addition after recent research, it is possible to format grpc messages into json format, but you need to take into account that the proto notation is different from json. In this case, the data will be in a more readable form, but not entirely true in terms of some encodings.

Example problem with ``bytes`` data type
```proto
message Response {
    bytes request_content = 1;
}
```

```java
new Response.newBuilder().setRequestContent(ByteString.fromUtf8("xml_request")).build().toString()
//would be 'request_content: ''xml_request" '
```

```java
JsonFormat.printer().print(new Response.newBuilder().setRequestContent(ByteString.fromUtf8("xml_request")).build())
//would be
//{
//    "request_content": "eG1sX3J1cXV1c3Q="
//}
```

So far, apart from problems with displaying Bytes, I have not found. Do we need pretty formatting, but with issues with bytes?

**Comment by baev** [comment]
Jackson uses base64 encoding when storing bytes. And what exactly is the problem?

**Comment by a-simeshin** [comment]
> Jackson uses base64 encoding when storing bytes. And what exactly is the problem?

Problem with readability and possible questions, like why I see one value in log.info (), but in the Allure report I see a completely different value, but only for ``bytes`` value.

Then if there is no objection, then I will add formatting to the pull request.

File: allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java (hunk 4 above)
Comment (a-simeshin) [reply allure-grpc/src/main/java/io/qameta/allure/grpc/CustomForwardingClientCall.java]: After many attempts, I still haven't found a working synchronous solution.

File: allure-grpc/build.gradle.kts (hunk 2 above)
Comment (a-simeshin) [reply allure-grpc/build.gradle.kts:20]: Also added dependencies for pretty formatting

**Comment by a-simeshin** [comment]
Added nice formatting for requests and responses, also added handling of the situation when the server sends many responses in for single request.

**Comment by a-simeshin** [comment]
Better one - #655

**Pull request closed without merging**